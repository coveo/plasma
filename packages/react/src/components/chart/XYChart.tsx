import * as d3 from 'd3';
import {FunctionComponent, createContext} from 'react';
import * as _ from 'underscore';

import {ChartBaseProps} from './ChartContainer';
import {ChartUtils} from './ChartUtils';

export interface XYPoint {
    x: number;
    y: number;
}

export interface XYSerie {
    label: string;
    data: XYPoint[];
}

export type Padding = {top?: number; right?: number; bottom?: number; left?: number};

export interface XYChartProps extends ChartBaseProps {
    series: XYSerie[];
    padding?: number | Padding;
    colorPattern?: string[];
    color?: (serie: number, colorPattern: string[], point?: XYPoint) => string;
    xFormat?: (value: number) => string | number;
    yFormat?: (value: number) => string | number;
    xTicksCount?: number;
    yTicksCount?: number;
}

export interface XYChartContextProps {
    width: number;
    height: number;
    xDomain: [number, number];
    yDomain: [number, number];
    series: XYSerie[];
    colorPattern: string[];
    xScale: d3.scale.Ordinal<number, number>;
    yScale: d3.scale.Linear<number, number>;
    xFormat: (value: number) => string | number;
    yFormat: (value: number) => string | number;
    xTicksCount: number;
    yTicksCount: number;
    color?: (serie: number, colorPattern: string[], point?: XYPoint) => string;
}

export const XYChartContext = createContext<XYChartContextProps>({
    width: 0,
    height: 0,
    xDomain: [0, 0],
    yDomain: [0, 0],
    series: [],
    colorPattern: [],
    xScale: d3.scale.ordinal<number, number>(),
    yScale: d3.scale.linear(),
    xFormat: _.identity,
    yFormat: _.identity,
    xTicksCount: 10,
    yTicksCount: 10,
});

const getDateChartColorPattern = (numOfColors: number) => {
    const colorPattern: string[] = [
        'var(--deprecated-azure)',
        'var(--deprecated-turquoise)',
        'var(--deprecated-anakiwa)',
        'var(--deprecated-nepal)',
    ];
    return numOfColors > 2 ? ['var(--deprecated-blue-8)', ...colorPattern] : colorPattern;
};
const defaultPadding = 10;

/**
 * @deprecated Use Mantine instead
 */
export const XYChart: FunctionComponent<XYChartProps> = ({
    width,
    height,
    series,
    colorPattern = [],
    color,
    padding = defaultPadding,
    xFormat = _.identity,
    yFormat = _.identity,
    xTicksCount,
    yTicksCount,
    children,
}) => {
    const xValues = ChartUtils.getXValues(series);
    const xDomain: [number, number] = [Math.min(...xValues), Math.max(...xValues)];

    const yValues = ChartUtils.getYValues(series);
    const yDomain: [number, number] = [Math.min(0, ...yValues), Math.max(...yValues)];

    colorPattern = colorPattern.length ? colorPattern : getDateChartColorPattern(series.length);
    color = color || ((serie: number, pattern: string[]) => pattern[serie]);

    const chartPadding: Padding = _.isNumber(padding)
        ? {top: padding, right: padding, bottom: padding, left: padding}
        : _.defaults(padding, {
              top: defaultPadding,
              right: defaultPadding,
              bottom: defaultPadding,
              left: defaultPadding,
          });

    width = width - chartPadding.left - chartPadding.right;
    height = height - chartPadding.top - chartPadding.bottom;

    const xScale = d3.scale.ordinal<number, number>().domain(xValues).rangePoints([0, width]);

    const yScale = d3.scale.linear<number, number>().domain(yDomain).range([height, 0]);

    return (
        <XYChartContext.Provider
            value={{
                width,
                height,
                xDomain,
                yDomain,
                xScale,
                yScale,
                xFormat,
                yFormat,
                color,
                colorPattern,
                series,
                xTicksCount: xTicksCount || xValues.length / 2,
                yTicksCount: yTicksCount || yValues.length / 2,
            }}
        >
            <svg width={width} height={height}>
                <g transform={`translate(${chartPadding.left},${chartPadding.top})`}>{children}</g>
            </svg>
        </XYChartContext.Provider>
    );
};
XYChart.displayName = 'XYChart';
