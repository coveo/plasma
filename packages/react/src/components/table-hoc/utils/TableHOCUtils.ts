import * as _ from 'underscore';

import {isValidElement} from 'react';
import {PlasmaState} from '../../../PlasmaState';
import {DatePickerSelectors} from '../../datePicker/DatePickerSelectors';
import {IFilterState} from '../../filterBox/FilterBoxReducers';
import {FlatSelectSelectors} from '../../flatSelect/FlatSelectSelectors';
import {IListBoxState} from '../../listBox/ListBoxReducers';
import {PaginationUtils} from '../../pagination/PaginationUtils';
import {ITableRowOwnProps} from '../TableRowConnected';
import {ITableWithSortState} from '../reducers/TableWithSortReducers';

export interface ITableHOCPredicateValue {
    id: string;
    value: string;
}

export interface ITableHOCCompositeState {
    sortKey?: string;
    sortAscending?: boolean;
    perPage?: number;
    pageNb?: number;
    filter?: string;
    predicates?: ITableHOCPredicateValue[];
    dateLimits?: [Date, Date?];
}

const PREDICATE_SEPARATOR = '--';

const getCompositeState = (id: string, state: PlasmaState): ITableHOCCompositeState => {
    const tableSort: ITableWithSortState = _.find(
        state.tableHOCHeader,
        (v: ITableWithSortState) => v.tableId === id && _.isBoolean(v.isAsc),
    );

    const perPage =
        parseInt(
            FlatSelectSelectors.getSelectedOptionId(state, {
                id: PaginationUtils.getPaginationPerPageId(id),
            }),
            10,
        ) || undefined;

    const filter: IFilterState = _.findWhere(state.filters, {id});
    const predicates = getTablePredicates(id, state);

    return {
        // predicates
        predicates,

        // sort
        sortKey: tableSort && tableSort.id,
        sortAscending: (tableSort && tableSort.isAsc) || null,

        // pagination
        perPage,
        pageNb: _.findWhere(state.paginationComposite, {id: getPaginationId(id)})?.pageNb,

        // filter
        filter: filter && filter.filterText,

        // date
        dateLimits: DatePickerSelectors.getDatePickerLimits(state, {id}),
    };
};

/** e.g. tableId--componentId => tableId */
const getTableIdFromPredicateId = (predicateId: string) => predicateId.split(PREDICATE_SEPARATOR)[0];

/** e.g. tableId--componentId => componentId */
const getComponentIdFromPredicateId = (predicateId: string) => predicateId.split(PREDICATE_SEPARATOR)[1];

const getPredicateId = (tableId: string, componentId: string) => `${tableId}${PREDICATE_SEPARATOR}${componentId}`;

const getPredicateIds = (tableId: string, state: PlasmaState): string[] =>
    _.chain(state.listBoxes)
        .filter(filterTablePredicate.bind(null, tableId))
        .pluck('id')
        .map(getComponentIdFromPredicateId)
        .value();

const getPaginationId = (tableId: string) => `pagination-${tableId}`;

const getTablePredicates = (tableId: string, state: PlasmaState): ITableHOCPredicateValue[] =>
    _.chain(state.listBoxes)
        .filter(filterTablePredicate.bind(null, tableId))
        .filter((list: IListBoxState) => list.selected && list.selected[0] !== '')
        .map((list: IListBoxState) => ({id: getComponentIdFromPredicateId(list.id), value: list.selected[0]}))
        .value();

const getDatePickerId = (tableId: string) => `${tableId}-date-range`;

export const isRowCollapsible = (props: ITableRowOwnProps): boolean =>
    props.collapsible &&
    (isValidElement(props.collapsible.content) ||
        _.isString(props.collapsible.content) ||
        _.isNull(props.collapsible.content));

export const TableHOCUtils = {
    getCompositeState,
    getPredicateId,
    getPredicateIds,
    getPaginationId,
    getTableIdFromPredicateId,
    getDatePickerId,
};

const filterTablePredicate = (tableId: string, {id}: IListBoxState): boolean => new RegExp(`^${tableId}(.+)`).test(id);
