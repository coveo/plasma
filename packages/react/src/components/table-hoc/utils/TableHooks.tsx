import {CSSProperties, useRef, useState, useLayoutEffect} from 'react';

export const useFixedWidthWhileLoading = (isLoading: boolean) => {
    const tableHeaderRef = useRef<HTMLTableHeaderCellElement>();
    const [columnWidth, setColumnWidth] = useState<number>();

    useLayoutEffect(() => {
        if (tableHeaderRef.current && !isLoading) {
            setColumnWidth(tableHeaderRef.current.offsetWidth);
        }
    }, [isLoading]);

    const style: CSSProperties = isLoading
        ? {
              width: `${columnWidth}px`,
          }
        : {};

    return {style, tableHeaderRef};
};
