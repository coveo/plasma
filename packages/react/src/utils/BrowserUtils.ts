// https://stackoverflow.com/questions/21825157/internet-explorer-11-detection#answer-22242528
const isIE = () =>
    typeof window !== 'undefined' &&
    (navigator.userAgent.indexOf('MSIE') !== -1 || navigator.appVersion.indexOf('Trident/') > -1);

export const BrowserUtils = {
    isIE,
};
