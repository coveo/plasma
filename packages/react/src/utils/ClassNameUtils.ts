import * as _ from 'underscore';
export type ClassNameTypes = string | string[] | {[key: string]: boolean};

/**
 * Type to use when typing a property or variable that will be used with
 * the "classname" library (https://www.npmjs.com/package/classnames)
 */
export type IClassName = ClassNameTypes | ClassNameTypes[];

export interface IAdditionalClass {
    className: string;
    condition?: (data: any) => boolean;
}

export const getAdditionalClasses = (additionalClasses: IAdditionalClass[], dataToUse: any): IClassName =>
    _.map(additionalClasses, (additionalClass: IAdditionalClass) =>
        additionalClass.condition
            ? {
                  [additionalClass.className]: additionalClass.condition(dataToUse),
              }
            : additionalClass.className,
    );
