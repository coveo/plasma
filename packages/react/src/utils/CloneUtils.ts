export const deepClone = (object: any) => JSON.parse(JSON.stringify(object));
