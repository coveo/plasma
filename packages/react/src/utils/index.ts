export * from './BrowserUtils';
export * from './ClassNameUtils';
export * from './CloneUtils';
export * from './ComponentUtils';
export * from './DataStructuresUtils';
export * from './DateUtils';
export * from './FalsyValuesUtils';
export * from './FilterUtils';
export * from './HocUtils';
export * from './InputUtils';
export * from './JSONUtils';
export * from './JSXUtils';
export * from './PropsToOmitUtils';
export * from './RandomUtils';
export * from './ReduxUtils';
export * from './StringUtils';
export * from './TooltipUtils';
export * from './TypesUtils';
export * from './UrlUtils';
export * from './UUID';
export * from './ValidComponentChildren';
export * from './TypeUtils';
