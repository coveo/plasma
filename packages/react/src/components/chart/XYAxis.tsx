import * as d3 from 'd3';
import {FunctionComponent, useContext} from 'react';
import * as _ from 'underscore';

import {XYChartContext} from './XYChart';

export interface AxisProps {
    show?: boolean;
    tickSize?: number;
    tickTextSize?: number;
    size?: number;
    innerPadding?: number;
}

export interface XYAxisProps {
    x?: AxisProps;
    y?: AxisProps;
}

const withDefaultConfig = (props: Partial<AxisProps> = {}): AxisProps =>
    _.defaults(props, {
        tickSize: props.show !== false ? 7 : 0,
        tickTextSize: 0,
        size: props.show !== false ? 40 : 0,
        innerPadding: 0,
        show: true,
        format: _.identity,
    });

/**
 * @deprecated Use Mantine instead
 */
export const XYAxis: FunctionComponent<XYAxisProps> = ({x, y, children}) => {
    const context = useContext(XYChartContext);
    const {xDomain, yDomain, xScale, yScale, xFormat, yFormat, width, height, xTicksCount, yTicksCount} = context;

    const xAxis = withDefaultConfig(x);
    const yAxis = withDefaultConfig(y);

    const yInnerPadding = yAxis.innerPadding || yAxis.tickTextSize / 2;
    const xInnerPadding = xAxis.innerPadding || xAxis.tickTextSize / 2;

    const newWidth = width - yAxis.size - 2 * xInnerPadding;
    const newHeight = height - xAxis.size - 2 * yInnerPadding;

    const newXScale = xScale.rangePoints([0, newWidth]);
    const newYScale = yScale.range([newHeight, 0]);

    const minX = newXScale(xDomain[0]);
    const maxX = newXScale(xDomain[1]);
    const minY = newYScale(yDomain[0]);
    const maxY = newYScale(yDomain[1]);

    const yTicks = newYScale
        .ticks(yAxis.tickTextSize === 0 ? yTicksCount : Math.floor(newHeight / yAxis.tickTextSize))
        .map((tick: number) => (
            <g
                key={`y-axis-tick-${tick}`}
                className="y-axis-tick"
                transform={`translate(${maxX + 2 * xInnerPadding},${newYScale(tick) + yInnerPadding})`}
            >
                <line stroke="black" x1="0" x2={yAxis.tickSize} />
                <text textAnchor="start" x={yAxis.tickSize + 5} y={5}>
                    {yFormat(tick)}
                </text>
            </g>
        ));

    const xLinearScale = d3.scale.linear().range(newXScale.range()).domain(xDomain);
    const xNumberOfTicks = xLinearScale.ticks(
        xAxis.tickTextSize === 0 ? xTicksCount : Math.floor(newWidth / xAxis.tickTextSize)
    ).length;
    const ticks = _.chunk(newXScale.domain(), Math.floor(newXScale.domain().length / xNumberOfTicks))
        // filter out the last array if it contains only one value and other arrays are bigger
        .filter(
            (arr: number[], index: number, arrays: number[][]) =>
                index < arrays.length - 1 || arr.length > 1 || arrays[0].length === 1
        )
        // get the element in the middle of the chunk
        .map((values: number[]) => values[Math.floor((values.length - 1) / 2)]);

    const xTicks = newXScale.domain().map((tick: number) => {
        const text = _.contains(ticks.length === 0 ? newXScale.domain() : ticks, tick) && xFormat(tick);
        const textX = newXScale(tick);
        return (
            <g
                key={`x-axis-tick-${tick}`}
                transform={`translate(${textX + xInnerPadding},${minY + 2 * yInnerPadding})`}
                className="x-axis-tick"
            >
                <line stroke="black" y1="0" y2={xAxis.tickSize} />
                {text && (
                    <text textAnchor="middle" y={xAxis.tickSize + 15}>
                        {text}
                    </text>
                )}
            </g>
        );
    });

    return (
        <>
            <XYChartContext.Provider
                value={{
                    ...context,
                    xScale: newXScale,
                    yScale: newYScale,
                    height: minY,
                    width: maxX,
                }}
            >
                <g transform={`translate(${xInnerPadding},${yInnerPadding})`}>{children}</g>
            </XYChartContext.Provider>
            {yAxis.show && (
                <g className="y-axis">
                    <line
                        className="axis-line"
                        stroke="black"
                        x1={maxX + 2 * xInnerPadding}
                        x2={maxX + 2 * xInnerPadding}
                        y1={minY + 2 * yInnerPadding}
                        y2={maxY}
                    />
                    {yTicks}
                </g>
            )}
            {xAxis.show && (
                <g className="x-axis">
                    <line
                        className="axis-line"
                        stroke="black"
                        x1={minX}
                        x2={maxX + 2 * xInnerPadding}
                        y1={minY + 2 * yInnerPadding}
                        y2={minY + 2 * yInnerPadding}
                    />
                    {xTicks}
                </g>
            )}
        </>
    );
};
XYAxis.displayName = 'XYAxis';
