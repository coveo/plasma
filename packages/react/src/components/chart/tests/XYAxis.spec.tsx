import {render, screen} from '../../../../jest/utils';

import {XYAxis} from '../XYAxis';
import {XYChartContext} from '../XYChart';
import {XYChartContextMock} from './XYChartContextMock';

describe('<XYAxis />', () => {
    it('should render as many x ticks as there are points in the scale from the context', () => {
        render(
            <XYChartContext.Provider value={XYChartContextMock}>
                <svg role="application" width={XYChartContextMock.width} height={XYChartContextMock.height}>
                    <XYAxis x={{show: true}} y={{show: false}} />
                </svg>
            </XYChartContext.Provider>,
        );

        expect(screen.getByRole('application')).toMatchSnapshot();
    });

    it('should render as many y ticks as defined in the context', () => {
        render(
            <XYChartContext.Provider value={XYChartContextMock}>
                <svg role="application" width={XYChartContextMock.width} height={XYChartContextMock.height}>
                    <XYAxis x={{show: false}} y={{show: true}} />
                </svg>
            </XYChartContext.Provider>,
        );

        expect(screen.getByRole('application')).toMatchSnapshot();
    });
});
