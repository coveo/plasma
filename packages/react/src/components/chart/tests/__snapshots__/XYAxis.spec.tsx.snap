// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XYAxis /> should render as many x ticks as there are points in the scale from the context 1`] = `
<svg
  height="50"
  role="application"
  width="10"
>
  <g
    transform="translate(0,0)"
  />
  <g
    class="x-axis"
  >
    <line
      class="axis-line"
      stroke="black"
      x1="0"
      x2="10"
      y1="10"
      y2="10"
    />
    <g
      class="x-axis-tick"
      transform="translate(0,10)"
    >
      <line
        stroke="black"
        y1="0"
        y2="7"
      />
      0
    </g>
    <g
      class="x-axis-tick"
      transform="translate(2,10)"
    >
      <line
        stroke="black"
        y1="0"
        y2="7"
      />
    </g>
    <g
      class="x-axis-tick"
      transform="translate(4,10)"
    >
      <line
        stroke="black"
        y1="0"
        y2="7"
      />
      <text
        text-anchor="middle"
        y="22"
      >
        2
      </text>
    </g>
    <g
      class="x-axis-tick"
      transform="translate(6,10)"
    >
      <line
        stroke="black"
        y1="0"
        y2="7"
      />
    </g>
    <g
      class="x-axis-tick"
      transform="translate(8,10)"
    >
      <line
        stroke="black"
        y1="0"
        y2="7"
      />
      <text
        text-anchor="middle"
        y="22"
      >
        4
      </text>
    </g>
    <g
      class="x-axis-tick"
      transform="translate(10,10)"
    >
      <line
        stroke="black"
        y1="0"
        y2="7"
      />
    </g>
  </g>
</svg>
`;

exports[`<XYAxis /> should render as many y ticks as defined in the context 1`] = `
<svg
  height="50"
  role="application"
  width="10"
>
  <g
    transform="translate(0,0)"
  />
  <g
    class="y-axis"
  >
    <line
      class="axis-line"
      stroke="black"
      x1="-30"
      x2="-30"
      y1="50"
      y2="0"
    />
    <g
      class="y-axis-tick"
      transform="translate(-30,50)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        100
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,45)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        90
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,40)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        80
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,35)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        70
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,30)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        60
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,25)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        50
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,20)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        40
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,15)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        30
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,10)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        20
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,5)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        10
      </text>
    </g>
    <g
      class="y-axis-tick"
      transform="translate(-30,0)"
    >
      <line
        stroke="black"
        x1="0"
        x2="7"
      />
      <text
        text-anchor="start"
        x="12"
        y="5"
      >
        0
      </text>
    </g>
  </g>
</svg>
`;
